examples/json-extension/client/src/extension.ts

```diff
@@ -20,80 +20,92 @@
+            { scheme: "file", language: "json" },
+            { scheme: "untitled", language: "json" },
+        ],
+        outputChannelName: "[pygls] JsonLanguageServer",
+        synchronize: {
+            // Notify the server about file changes to '.clientrc files contain in the workspace
+            fileEvents: workspace.createFileSystemWatcher("**/.clientrc"),
+        },
+    };
 }
 
 function isStartedInDebugMode(): boolean {
-  return process.env.VSCODE_DEBUG_MODE === "true";
+    return process.env.VSCODE_DEBUG_MODE === "true";
 }
 
-function startLangServerTCP(addr: number): LanguageClient {
-  const serverOptions: ServerOptions = () => {
-    return new Promise((resolve, reject) => {
-      const clientSocket = new net.Socket();
-      clientSocket.connect(addr, "127.0.0.1", () => {
-        resolve({
-          reader: clientSocket,
-          writer: clientSocket,
+function startLangServerTCP(port: number): LanguageClient {
+    const serverOptions: ServerOptions = () => {
+        return new Promise((resolve, reject) => {
+            const clientSocket = new net.Socket();
+            clientSocket.connect(port, "127.0.0.1", () => {
+                resolve({
+                    reader: clientSocket,
+                    writer: clientSocket,
+                });
+            });
         });
-      });
-    });
-  };
+    };
 
-  return new LanguageClient(`tcp lang server (port ${addr})`, serverOptions, getClientOptions());
+    return new LanguageClient(`tcp lang server (port ${port})`, serverOptions, getClientOptions());
 }
 
 function startLangServer(
-  command: string, args: string[], cwd: string,
+    command: string, args: string[], cwd: string,
 ): LanguageClient {
-  const serverOptions: ServerOptions = {
-    args,
-    command,
-    options: { cwd },
-  };
+    const serverOptions: ServerOptions = {
+        args,
+        command,
+        options: { cwd },
+    };
 
-  return new LanguageClient(command, serverOptions, getClientOptions());
+    return new LanguageClient(command, serverOptions, getClientOptions());
+}
+
+function getConfig(key: string): WorkspaceConfiguration {
+    return workspace.getConfiguration(key, editor ? editor.document.uri : null);
+}
+
+function getPython(): string {
+    return getConfig("python").get<string>("pythonPath");
 }
 
 export function activate(context: ExtensionContext) {
-  if (isStartedInDebugMode()) {
-    // Development - Run the server manually
-    client = startLangServerTCP(2087);
-  } else {
-    // Production - Client is going to run the server (for use within `.vsix` package)
-    const cwd = path.join(__dirname, "../");
-    const pythonPath = workspace.getConfiguration("python").get<string>("pythonPath");
+    const debugPort = process.env.DebugTCP;
+    const config = getConfig("jsonServer");
+    if (debugPort) {
+        // Debug port: server should be running externally on specified port.
+        client = startLangServerTCP(Number(debugPort));
+    } else {
```

Comment:
I think it would be easier for user to try it out if you add `serverPath` and `serverArgs` to environment variables [here](https://github.com/openlawlibrary/pygls/blob/master/examples/json-extension/.vscode/launch.json#L18) and push that file.
Then, instead doing `config.get<string>("serverPath");` you can get config values with `process.env.SERVER_EXECUTABLE_PATH` and `process.env.SERVER_ARGS`.

Finally, we should update docs and say that [this](https://github.com/openlawlibrary/pygls/blob/master/examples/json-extension/.vscode/launch.json#L19) environment variable needs to be _False_ in order to run pre-compiled server.

```typescript
if (isStartedInDebugMode()) {
  // Development - Run the server manually
  client = startLangServerTCP(2087);
} else {
  // Client is going to run the server
  const cwd = path.join(__dirname, "../");

  const serverPath = process.env.SERVER_EXECUTABLE_PATH;
  const serverArgs = process.env.SERVER_ARGS;

  if (serverPath) {
    // Run pre-compiled server
    client = startLangServer(serverPath, serverArgs, cwd);
  } else {
    // Run server from source
    client = startLangServer(pythonPath, ["-m", "server"], cwd);
  }
}
```

@TiemenSch What are your thoughts here?

Reply:
Personally, I really like the VS Code settings approach because this makes it easier for users of the extension to point it to any executable/args combination even when not debugging. These settings are also neatly documented out of the box in VS Code's settings menu's.

 It also doesn't require you to clog the environment variables any more than necessary :).

Reply:
That's fine, I agree mostly, but I am usually running the example without opening any folder and I don't want to add those settings under user-settings.

You can set those variables depending on OS:

```json
...
 "env": {
                "VSCODE_DEBUG_MODE": "true"
            },
            "osx": {
                "env": {
                    "SERVER_EXECUTABLE_PATH": "build/json_server.dist/json_server",
                }
            },
            "linux": {
                "env": {
                    "SERVER_EXECUTABLE_PATH": "build/json_server.dist/json_server",
                }
            },
            "windows": {
                "env": {
                    "SERVER_EXECUTABLE_PATH": "build/json_server.dist/json_server.exe",
                }
            },
...
```

and append them on `cwd` path. This is just an example and for production, someone will need to figure out how to package and pre-compiled server, depending on platform.

That's just from my perspective when trying to run example and review a PR. :)

Reply:
Ah, we could look for both env variables and settings (in what order?) to account for this.

I just open a folder with the right workspacesettings depending on what use-case I want to debug ;). It's not that I don't trust myself changing my system's path, but rather that I would not want end-users to have to do so.

Reply:
You could check settings primary and then env vars, but I guess you won't need to do this.
In case that I am suggesting, you (and any other user) won't need to change and nor setup anything. https://github.com/openlawlibrary/pygls/pull/76#discussion_r327674536 is a content of `launch.json` file under `json-example/.vsode` directory and has relative path to the _json-example_ directory. The only thing that they need to do is to install and compile server and then launch the client.